lib/service/service.go

```diff
@@ -494,6 +506,50 @@ func (process *TeleportProcess) SetExpectedInstanceRole(role types.SystemRole, e
 	process.instanceRoles[role] = eventName
 }
 
+// RegisterDynamicInstanceRole will set a dynamic instance role for the process. This will set the role
+// and update the inventory.
+func (process *TeleportProcess) RegisterDynamicInstanceRole(role types.SystemRole) {
+	process.dynamicInstanceRolesMu.Lock()
+	defer process.dynamicInstanceRolesMu.Unlock()
+	process.log.Infof("Registering dynamic role: %v", role)
+	process.dynamicInstanceRoles[role]++
+
+	if process.dynamicInstanceRoles[role] == 1 {
+		process.updateDownstreamPingHandler(process.Config)
+	}
+}
+
+// UnregisterDynamicInstanceRole will unregister a dynamic instance role for the process. This will update
+// the inventory if needed.
+func (process *TeleportProcess) UnregisterDynamicInstanceRole(role types.SystemRole) error {
+	process.dynamicInstanceRolesMu.Lock()
+	defer process.dynamicInstanceRolesMu.Unlock()
+
+	// There's no corresponding count for this dynamic instance role, so we'll return an error.
+	if process.dynamicInstanceRoles[role] == 0 {
+		return trace.BadParameter("unregister dynamic instance role called without corresponding register dynamic instance call")
+	}
+
+	process.log.Infof("Unregistering dynamic role: %v", role)
+	process.dynamicInstanceRoles[role]--
+
+	if process.dynamicInstanceRoles[role] == 0 {
+		process.updateDownstreamPingHandler(process.Config)
+	}
+
+	return nil
+}
+
+func (process *TeleportProcess) getDynamicInstanceRoles() []types.SystemRole {
+	out := make([]types.SystemRole, 0, len(process.dynamicInstanceRoles))
+	for role, count := range process.dynamicInstanceRoles {
```

Comment:
no locking for `process.dynamicInstanceRoles`? this could be mutated concurrently, so this should be done under the related lock.